### ISSUE TITLE
Update multiplayer API to prevent fog/shroud/ambush disclosure

### ISSUE THREAD

**Issue by renom** [wesnoth/wesnoth#5120]
The client should receive only the visible part of the map to avoid possible cheating. I know it's hard to implement, and it has a few drawbacks, but I don't think it's a good reason to never fix the issue.

**Comment by gfgtdf** [comment]
This is currently not possible, the way the multiplayer server works is that is doesn't calculate anything game relatred and just forwards data send from one clients to the other clients. Changing this currently require reimplementing quite large parts of the code which we currentyl cannot afford, furthermore it would  break quite a few addons which we also don't want.

**Comment by soliton-** [comment]
It is questionable whether it is really a good idea to make such a drastic design change.

You should list the drawbacks and advantages and explain why you think this is still worth it in your opinion.

**Comment by renom** [comment]
@gfgtdf could you explain why it would break some addons? If the server calculated the game state itself, none of the addons would be broken as the server would do the same job as the game client does in the current implementation. What is needed is to separate the game engine logic from the game client so that it would be possible to use it in the game client (for single player) as well as in the server (for multiplayer). Then it would be needed to update the network API in order that the server sends only the visible parts of the map to the clients.
@soliton- I think the main drawback is it would be difficult to implement proper logic for ambush (the game client should request the ambush info from the server when the unit moves). So the undo-move logic should be also redesigned so that the server would decide whether it is possible to undo a move or not.
The advantage is that the game would become more suitable for tournaments and ranked games. I think it'd even be beneficial for regular games as any player would be confident that nobody is cheating.

**Comment by soliton-** [comment]
You seem to not see moving the game logic to the server side as an issue at all. How do you suggest multiplayer games are hosted in the future?

**Comment by gfgtdf** [comment]
> could you explain why it would break some addons?

Because you cannot move all umc code execution easily to the server, think for example of the lua callbacks that generate the information on right side panel, you need to update this quickly (and so far not requiring a fast connection was a nice feature of wesnoth). And  currently the addons code that implemented these accesses the gamestate variables freely. There are other parts parts of the api like this for example the show_dialog functions etc.

**Comment by renom** [comment]
Well, I think it's possible to implement alternative server API which wouldn't support all the features that the game offers. I mean, wesnoth would still support the current mode (P2P), which implies low server involvement, and at the same time it would be possible to use a centralized solution, which would calculate the game state itself and game clients wouldn't receive ambush/fog/shroud info. Also it'd become impossible to save the game from the client, so the server would rehost games itself (it would prevent the save-file editing cheating as well).
Most of the tournaments (including ladder) are played with the default era, and on simple scenarios, so there'd be no need to support Lua and most of the WML features (IIRC the default era doesn't use most of the modding features). So I think it'd be fairly easy to support such basic feature set in the centralized solution.
Anyway, the main difficulty would be to make the server able to calculate the game stuff. Do you think it is possible to reuse already implemented code for this?

**Comment by gfgtdf** [comment]
> Well, I think it's possible to implement alternative server API

No maintaining two differnt server modes is even more work. its also more work for umc authros to understand.

> Anyway, the main difficulty would be to make the server able to calculate the game stuff. Do you think it is possible to reuse already implemented code for this?

some parts could clearly be reused.


I think if you really want this, your only chance is to work on this yourself, you can either try to work on the main wesnoths cpp files, or work on fabis new wesnoth-löve project which iirc also implements a similar server-client structure.

**Comment by renom** [comment]
Hey @gfgtdf, thanks for the info, I didn't know that there's another wesnoth port is in development. Does this mean that the haldric project was abandoned?

**Comment by gfgtdf** [comment]
> Hey @gfgtdf, thanks for the info, I didn't know that another wesnoth port is in development. Does this mean that the haldric project was abandoned?

No its just another approach by another developer. I'm not yet sure hwo far exactly he is, you can find his topic on the forum.

**Comment by stevecotton** [comment]
Another possible approach: running a hotseat game on a server, and then the players connect to it via a remote-desktop protocol such as VNC. There would be a lot of questions about the security of that setup, but it might be something that has already been solved for another game.

A quick web search found discussions where such a setup has been considered before, although these don't give answers:
* https://gaming.stackexchange.com/questions/123182/civ-5-over-remote-session
* https://www.gog.com/forum/general/how_to_play_a_hotseat_dosgame_online

**Comment by AI0867** [comment]
That would be an impressively resource-intensive way to solve the problem. Instead of a server that keeps a log and mostly just passes messages from the players to each other, you would need a full wesnoth client, possibly one for each player, and stream rendered graphics.

While our server is no slouch, I'm pretty sure it would not be able to handle being subjected to that type of load.

**Comment by renom** [comment]
As both clones aren't actively developed recently I suggest reconsidering moving game logic to the server for mp. I think we need to split up the APIs into the following two groups: the ones that handle sensitive information about the other player sides and the ones that does not. I believe the Lua APIs belong to the first group. So most APIs would be processed on the server, but some would be processed on the client. What do you think?

**Comment by stevecotton** [comment]
> [hotseat via remote-desktop protocol] would be an impressively resource-intensive way to solve the problem.

Agreed, I should have clarified that I'd expect someone to rent a server in same way that FPS Deathmatch clans rent servers. However, it doesn't sound that resource-intensive for a world that already has video-streaming, music-streaming, Twitch and even gaming-as-a-service.

**Comment by stevecotton** [comment]
> I believe the Lua APIs belong to the second group.

The Lua APIs include many way to find units. There's the direct wesnoth.get_units(), there's access to the pathfinder (which must take account of ambushers) and plenty of less direct ways to work out where units are.

**Comment by AI0867** [comment]
Any solution to this would pretty much require a ground-up redesign of the entire MP system.
The current system is essentially single-player except you also transmit all your moves and part of the time some other entity is sending the control inputs. Every wesnoth client knows and processes the same gamestate.

I see two possible approaches:

The first is to make the server smarter and have it do all the gamestate processing. It will need to observe the gamestate from each side's perspective and only send the parts that that side should know about. Unfortunately, this also means that every move made by a player will require a round-trip to the server to see if it succeeds, any revealed fog and what the results are. It would probably be preferable to move AI processing to the server too, though this would increase server load.

The second keeps the server in its current role, but instead has each side keep its information private as much as possible. Every move would now require a broadcast to every other player, and waiting for a response from each. I don't know if Zero Knowledge Proofs apply here, but something similar can probably be achieved.

In both cases though, the way the game processes multiplayer games would have to be changed significantly.
